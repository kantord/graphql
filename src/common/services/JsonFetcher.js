// @flow

import fetch from 'node-fetch';
import Logger from '../services/Logger';
import { ProxiedError } from './errors/ProxiedError';

export async function fetchJson(
  url: string,
  method: string = 'GET',
  options?: Object = {},
): Promise<any> {
  if (process.env.NODE_ENV === 'test') {
    throw new Error(
      `HttpRequest should never be called in test environment. Have you forgotten to mock "${url}" with fake data response?`,
    );
  }
  if (typeof url !== 'string') {
    throw new Error(
      `Relative API URL should be typeof 'string', '${typeof url}' given.`,
    );
  }

  const reqOptions = {
    ...options,
    method,
  };

  Logger.info(`${method} ${url}`);

  const response = await fetch(url, reqOptions);

  if (response.status >= 300) {
    throw new ProxiedError(response.statusText, response.status, url);
  }
  return response.json();
}
