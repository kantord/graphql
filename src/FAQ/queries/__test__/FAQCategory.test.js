// @flow

import { graphql, RestApiMock } from '../../../common/services/TestingTools';
import rootCategory from '../../datasets/rootCategory.json';
import categories from '../../datasets/categories.json';

describe('allFAQCategories', () => {
  beforeEach(() => {
    RestApiMock.onGet(
      'https://api.skypicker.com/knowledgebase/api/v1/categories',
    ).replyWithData(rootCategory);
    RestApiMock.onGet(
      'https://api.skypicker.com/knowledgebase/api/v1/categories/1',
    ).replyWithData(categories);
  });

  it('should return single FAQ category', async () => {
    const id = 'RkFRQ2F0ZWdvcnk6NDc='; // "Before the flight"
    const resultsQuery = `query FAQSubcategories($id: ID!) { 
      FAQCategory(language: en, id: $id) {
        id
        title
        subcategories {
          id
        }
        FAQs {
          id
          title
        }  
      }      
    }`;
    expect(await graphql(resultsQuery, { id })).toMatchSnapshot();
  });

  it('should return ancestors for FAQ category', async () => {
    const id = 'RkFRQ2F0ZWdvcnk6NDM='; // "Searching for flights"
    const resultsQuery = `query FAQSubcategories($id: ID!) { 
      FAQCategory(language: en, id: $id) {
        id
        title
        ancestors {
          id
          title
          subcategories {
            id
            title
          }
        }
        FAQs {
          id
          title
        }  
      }      
    }`;
    expect(await graphql(resultsQuery, { id })).toMatchSnapshot();
  });

  it('should return error for non-existing category', async () => {
    const id = 'RkFRQ2F0ZWdvcnk6NjY2'; // non-existing category #666
    const resultsQuery = `query FAQSubcategories($id: ID!) { 
      FAQCategory(language: en, id: $id) {
        id
      }        
    }`;
    expect(await graphql(resultsQuery, { id })).toMatchSnapshot();
  });
});
