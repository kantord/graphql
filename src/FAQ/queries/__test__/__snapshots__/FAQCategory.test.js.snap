// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allFAQCategories should return ancestors for FAQ category 1`] = `
Object {
  "data": Object {
    "FAQCategory": Object {
      "FAQs": Array [
        Object {
          "id": "RkFRQXJ0aWNsZTo5NQ==",
          "title": "What languages do you support?",
        },
        Object {
          "id": "RkFRQXJ0aWNsZTo5Mw==",
          "title": "What's the difference between low-cost and legacy airline carriers?",
        },
        Object {
          "id": "RkFRQXJ0aWNsZTo5NA==",
          "title": "How do I find the best flights for my budget?",
        },
        Object {
          "id": "RkFRQXJ0aWNsZToxMTE=",
          "title": "How do I use the mobile app?",
        },
      ],
      "ancestors": Array [
        Object {
          "id": "RkFRQ2F0ZWdvcnk6NDY=",
          "subcategories": Array [
            Object {
              "id": "RkFRQ2F0ZWdvcnk6NDM=",
              "title": "Searching for flights",
            },
            Object {
              "id": "RkFRQ2F0ZWdvcnk6MzA=",
              "title": "E-ticket",
            },
            Object {
              "id": "RkFRQ2F0ZWdvcnk6Mjc=",
              "title": "Booking a Flight",
            },
            Object {
              "id": "RkFRQ2F0ZWdvcnk6MzM=",
              "title": "Baggage",
            },
            Object {
              "id": "RkFRQ2F0ZWdvcnk6Mjg=",
              "title": "Payment Issues",
            },
          ],
          "title": "Booking",
        },
      ],
      "id": "RkFRQ2F0ZWdvcnk6NDM=",
      "title": "Searching for flights",
    },
  },
}
`;

exports[`allFAQCategories should return error for non-existing category 1`] = `
Object {
  "data": Object {
    "FAQCategory": null,
  },
  "errors": Array [
    [GraphQLError: No FAQ category found with ID RkFRQ2F0ZWdvcnk6NjY2],
  ],
}
`;

exports[`allFAQCategories should return single FAQ category 1`] = `
Object {
  "data": Object {
    "FAQCategory": Object {
      "FAQs": Array [
        Object {
          "id": "RkFRQXJ0aWNsZToxMzI=",
          "title": "How do I Refer a Friend?",
        },
        Object {
          "id": "RkFRQXJ0aWNsZTozOQ==",
          "title": "How do I know which terminals I will be using?",
        },
      ],
      "id": "RkFRQ2F0ZWdvcnk6NDc=",
      "subcategories": Array [
        Object {
          "id": "RkFRQ2F0ZWdvcnk6Mzc=",
        },
        Object {
          "id": "RkFRQ2F0ZWdvcnk6NDA=",
        },
        Object {
          "id": "RkFRQ2F0ZWdvcnk6MzU=",
        },
        Object {
          "id": "RkFRQ2F0ZWdvcnk6MzY=",
        },
        Object {
          "id": "RkFRQ2F0ZWdvcnk6MjY=",
        },
        Object {
          "id": "RkFRQ2F0ZWdvcnk6MzE=",
        },
        Object {
          "id": "RkFRQ2F0ZWdvcnk6NDI=",
        },
      ],
      "title": "Before the flight",
    },
  },
}
`;
